Fix 1409 chart workflow

mattroumaya opened the pull request:
## Overview
Fix #1312 
Fix #1437 


## Test Notes/Sample Code
<!--- Notes about testing or code to reproduce new functionality --->

Notes:

mattroumaya commented:
@jwildfire this is ready for you to check out. I think the most straightforward check would be to just run some cumulative snapshots using the new model and inspecting the `snapshot$lCharts` object:

```
snap_one <- Make_Snapshot(strAnalysisDate = "2022-01-01")
snap_two <- Make_Snapshot(strAnalysisDate = "2022-02-01", lPrevSnapshot = snap_one)
snap_three <- Make_Snapshot(strAnalysisDate = "2022-03-01", lPrevSnapshot = snap_two)
```

jwildfire commented on R/MakeKRICharts.R:
```diff
@@ -56,5 +59,28 @@ MakeKRICharts <- function(lData = NULL) {
+  # Continuous Charts -------------------------------------------------------
+    if (!is.null(lStackedSnapshots)) {
+
+      number_of_snapshots <- length(unique(lStackedSnapshots$rpt_site_kri_details$snapshot_date))
```
Could we just combine `lStackedSnapshots` and `lData`? Could the stacked version of `rpt_site_kri_details`  just (optionally) be added to lData? Is `stacked$rpt_site_kri_details` really needed? Will look a bit more ...

jwildfire commented on R/Widget_TimeSeries.R:
```diff
@@ -41,6 +41,48 @@ Widget_TimeSeries <- function(
+  # -- this is the data format expected by JS library {rbm-viz}
+  results <- results %>%
+    select(
+      "studyid" = "study_id",
+      "groupid" = "site_id",
+      "numerator",
+      "denominator",
+      "metric" = "kri_value",
+      "score" = "kri_score",
+      "flag" = "flag_value",
+      "gsm_analysis_date",
+      "snapshot_date"
```
Hmm. I wonder if we should refactor the assess functions to use this naming. I guess this rename is happening in make_snapshot now ...

jwildfire replied:
yeah, going back and forth between these naming conventions is definitely a pain and will be hard to mainta. Let's discuss how to clean this up.

jwildfire commented on R/util-MakeSnapshotHelpers.R:
```diff
@@ -310,26 +310,26 @@ MakeRptKriDetails <- function(lResults, status_site, meta_workflow, status_workf
       pt_data_dt = NA_character_
     ) %>%
     select("study_id",
```
Per comment above, all this manual renaming isn't ideal. Options: 
- capture all of this as a `mapping` object
- refactor the assessment functions to use the new naming
- update the data model to use the old naming 

cc: @samussiah

jwildfire commented on R/Widget_TimeSeries.R:
```diff
@@ -41,6 +41,48 @@ Widget_TimeSeries <- function(
     siteSelectLabelValue <- paste0("Highlighted ", siteSelectLabelValue, ": ")
   }
 
+  # rename results to account for rpt_* table refactor
```
Ideally the widgets will take snapshot outputs as inputs. it's not entirely clear how `results` here maps to the snapshot outputs. 

Need to take a look at this and think through how that will work in the new charting paradigm. Going to be closely related to how we handle the mapping between `assessment$lResults` and `snapshot$lData` as discussed above.

jwildfire reviewed (commented):
Adding my initial thoughts as a  series of comments for now. My main comment is that I would like to tighten up the data standards between our "Assessment" and "Snapshot" objects before release, and then update the naming conventions/parametrization in the charting/widget functions to very clearly tie to those data standards. Will look for some time to discuss.

mattroumaya commented:
@jwildfire - I think this one is ready for you to re-review. 

- added `lLabels` to each assessment, which could be specified in a YAML workflow if needed. 
- In `Make_Snapshot()`, the `dfConfig` object is overwritten by default from `lMeta$meta_workflow` based on `workflowid`, except for the thresholds which are saved as the actual thresholds used when the assessment/workflow was run. 

In general, this is slightly shaky since `Make_Snapshot()` is not yet entirely metadata-driven.

jwildfire commented on R/AE_Assess.R:
```diff
@@ -188,15 +199,8 @@ AE_Assess <- function(
-      strDenominator = "Days on Study",
-      vThreshold = vThreshold
-    )
+    lData$dfConfig <- purrr::map_df(lLabels, ~.x) %>%
+      mutate(thresholds = list(vThreshold))
```
Can we just not have `dfConfig`?

jwildfire reviewed (commented):
A very thorough review ...

jwildfire commented on R/Make_Snapshot.R:
```diff
@@ -144,21 +144,23 @@ Make_Snapshot <- function(
       if (x$bStatus) {
 
         if (!grepl("qtl", x$name)) {
+
+          dfWorkflow <- lMeta$meta_workflow %>% filter(workflowid == x$name) %>%
+            mutate(
+              thresholds = list(x$lResults$lData$dfConfig$thresholds)
+            )
+
```
Can we generate `lLabels` here instead?

jwildfire commented on R/Make_Snapshot.R:
```diff
@@ -144,21 +144,23 @@ Make_Snapshot <- function(
             lData = x$lResults$lData,
             lStackedSnapshots = SubsetStackedSnapshots(workflowid = x$name, lStackedSnapshots = lStackedSnapshots),
-            dfMetaWorkflow = lMeta$meta_workflow
+            dfWorkflow = dfWorkflow
```
```suggestion
            lLabels = lLabels
```

jwildfire commented on R/MakeKRICharts.R:
```diff
@@ -5,46 +5,46 @@
 #'
 #' @export
-MakeKRICharts <- function(strWorkflowId = NULL, lData = NULL, lStackedSnapshots = NULL, dfMetaWorkflow = NULL) {
+MakeKRICharts <- function(strWorkflowId = NULL, lData = NULL, dfWorkflow = NULL, lStackedSnapshots = NULL) {
```
```suggestion
MakeKRICharts <- function(strWorkflowId = NULL, lData = NULL, lLabels = NULL, lStackedSnapshots = NULL) {
```

This is really the key change I'm proposing. I think it makes more sense to pass a list of labels instead of a data.frame with one row.

mattroumaya commented:
To update `Make_Snapshot()` header (in progress):

```
a <- rbm_data_spec %>% 
  filter(System == "Gismo") %>% 
  split(.$Table) %>% 
  purrr::imap(
    function(x, y) {
      glue::glue("   - `{x$Column}` - {x$Description}")
    }
  )

b <- rbm_data_spec %>% 
  filter(System == "Gismo") %>% 
  split(.$Table) %>% 
  purrr::imap(~ glue::glue("- `{.y}`, with columns:"))

```

jwildfire commented on R/MakeKRICharts.R:
```diff
@@ -1,50 +1,51 @@
 #' @param lData `list` A list containing two data frames - dfSummary and dfBounds.
 #' @param lStackedSnapshots `list` A list returned by [gsm::Make_Snapshot()] containing flat files for snapshot/assessment data.
-#' @param dfWorkflow `data.frame` Workflow metadata. See [gsm::meta_workflow].
+#' @param lLabels `data.frame` Workflow metadata. See [gsm::meta_workflow].
```
```suggestion
#' @param lLabels `list` Workflow metadata. See [gsm::meta_workflow].
```

jwildfire commented on R/MakeKRICharts.R:
```diff
@@ -1,50 +1,51 @@
 #'
 #' @export
-MakeKRICharts <- function(strWorkflowId = NULL, lData = NULL, dfWorkflow = NULL, lStackedSnapshots = NULL) {
+MakeKRICharts <- function(strWorkflowId = NULL, lData = NULL, lLabels = NULL, lStackedSnapshots = NULL) {
```
remove workflowID?

jwildfire commented on R/MakeKRICharts.R:
```diff
@@ -1,50 +1,51 @@
+#' The function MakeKRICharts creates three different types of charts (scatter plot and two bar charts) using the gsm package.
 #'
 #' @param strWorkflowId `character` Name of workflow ID, e.g., 'kri0001'.
 #' @param lData `list` A list containing two data frames - dfSummary and dfBounds.
```
Consider sending 2 seprarate df

jwildfire commented on R/MakeKRICharts.R:
```diff
@@ -1,50 +1,51 @@
-            elementId = paste0(tolower(dfWorkflow$abbreviation), "AssessScatter"),
-            siteSelectLabelValue = dfWorkflow$group
+            elementId = paste0(tolower(lLabels$abbreviation), "AssessScatter"),
+            siteSelectLabelValue = lLabels$group
```
Just pull this from lLabels in the `results` param

mattroumaya commented:
- Pull data out of lists in `MakeKRICharts()` to make very obvious which datasets we need to create the visualization.
